### ISSUE TITLE
Add support for difftastic?

### ISSUE THREAD

**Issue by pvonmoradi** [jesseduffield/lazygit#1853]
[difftastic](https://github.com/wilfred/difftastic) is a diff tool which is [aware of the semantics](https://difftastic.wilfred.me.uk/introduction.html) of the code depending on their programming language.

I was wondering if it is possible to integrate it into lazygit as a custom diff tool.
I've tried adding it to lazygit config as a custom pager like this:
```
git:
  paging:
    colorArg: always
    pager: difft
```
but I guess this is not the correct way of doing it since it prints its `--help` in diff section.

**Comment by RaafatTurki** [comment]
same issue, would love to see difftastic supported

**Comment by jesseduffield** [comment]
It should be fairly easy to set up, though admittedly after a few minutes reading the [docs](https://difftastic.wilfred.me.uk/git.html) I haven't been able to get it working. Not sure if we need a code change on the lazygit side or if it's just a matter of coming up with the right command template in the config

**Comment by pvonmoradi** [comment]
> Not sure if we need a code change on the lazygit side or if it's just a matter of coming up with the right command template in the config

Is there a way to pass env. vars to `lazygit`? 
`GIT_EXTERNAL_DIFF='difft --display inline' git diff` uses difft (with paging) but `GIT_EXTERNAL_DIFF='difft --display inline' lazygit` does not.

btw, `difft` version 0.26 added a `--display` switch which accepts an `inline` value which may be more proper for lazygit since the side-by-side mode may not fit in the main pane.

**Comment by jesseduffield** [comment]
Lazygit invokes git as a subprocess, and for the pager we use the GIT_PAGER env var e.g. `GIT_PAGER=difft`. So perhaps we could conditionally use `GIT_EXTERNAL_DIFF` instead of `GIT_PAGER`?

**Comment by jesseduffield** [comment]
when I run `GIT_EXTERNAL_DIFF=difft git diff` in the CLI that works fine, but when I do `cmd.Env = append(cmd.Env, "GIT_EXTERNAL_DIFF=difft")` in `pkg/gui/pty.go:53` (`newPtyTask` function) it appears like the regular diff mode

**Comment by OlaoluwaM** [comment]
Any status on this?

**Comment by aruediger** [comment]
So the difference to other tools is that difftastic is an alternative diff tool while e.g. diff-so-fancy is implemented as a pager?

Would be nice if `git config --global diff.external difft` would be sufficient to convince lazygit to pick it up. Might be that some heuristic or an additional switch is needed to support `diff.external` as well as `core.pager`.

It looks like lazygit currently prevents this by [setting](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/diffing.go#L19) `-no-ext-diff`.

**Comment by MSDimos** [comment]
Any status update on this, pls?

**Comment by jesseduffield** [comment]
@aruediger we can make that --no-ext-diff thing configurable. Would that suffice? Happy to accept a PR for this

**Comment by PhotonQuantum** [comment]
@aruediger I tested your idea. Unfortunately, lazygit's hunk parsing algorithm is not compatible with difftastic's output format, even with inline display set. It results in something like this:

<img width="1237" alt="截屏2022-11-10 02 32 10" src="https://user-images.githubusercontent.com/18749973/200912350-73a818c5-028a-4560-a92b-3112884e6a62.png">

**Comment by aruediger** [comment]
@PhotonQuantum Bummer. (Was just about going to look into it myself.)

@jesseduffield How hard would it be to teach lazygit to deal with difftastic's chunk format?

**Comment by hacker-DOM** [comment]
Fwiw: it would be great to interactively switch between various differs / pagers. Most of the time I prefer `difftastic`, but it would be nice to switch over to `delta` without leaving lazygit.

**Comment by hacker-DOM** [comment]
People are Dming me to send them my config 😂 I was just pointing out that once we do figure out a way to integrate them, we shoud want to be able to switch on the fly. I use lazygit from neovim and the speed difference between being able to switch on the go and closing lazygit, editing my config, and relaunching it (or potentially all of neovim) is like 1000x. So no, I do not have it working.

**Comment by stefanhaller** [comment]
Here's a PR that adds support for difftastic: #2868. Testing and feedback welcome.

However, I have to say that I'm a little underwhelmed. There are a few cases occasionally where difftastic does a great job at rendering the diff; for example in go code when indentation changes, like [here](https://github.com/jesseduffield/lazygit/commit/917cfd758689#diff-3eb5426752dae525c92f1ecce2f7215d15bfdec91bd459e08f78574876583910L1477). However, in most cases like this I can get a very similar result from any other pager by simply hitting `<c-w>` to ignore whitespace. For the vast majority of my work, I still greatly prefer delta's output.

I suppose people working in other programming languages than I do might find it useful more often.

I agree with the comments above that suggest to add a keybinding to cycle through pagers. I don't want this badly enough myself to explore this any time soon, though.

**Comment by stefanhaller** [comment]
Closing this as #2868 is on master now.

**Comment by alex007sirois** [comment]
Do we know when this feature will be released? In other words, do we have an ETA on the next release?

**Comment by pookdeveloper** [comment]
same issue, would love to see difftastic supported please!

**Comment by stefanhaller** [comment]
@pookdeveloper This issue is closed, support for difftastic was added almost two years ago. See [docs](https://github.com/jesseduffield/lazygit/blob/master/docs/Custom_Pagers.md#using-external-diff-commands).

**Comment by ahmafi** [comment]
In the docs it's mentioned that we can cycle through pagers with `|` key. But I can't find this shortcut anywhere!

This is my `config.yml`:

```yml
git:
  pagers:
    - pager: delta --dark --paging=never
    - externalDiffCommand: difft --color=always
```

**Comment by gwennlbh** [comment]
> In the docs it's mentioned that we can cycle through pagers with `|` key. But I can't find this shortcut anywhere!
> 
> This is my `config.yml`:
> 
> git:
>   pagers:
>     - pager: delta --dark --paging=never
>     - externalDiffCommand: difft --color=always

I looked in the keybindings docs, there's an entry for "Cycle pagers" but the associated bind is empty: https://github.com/jesseduffield/lazygit/blob/master/docs/keybindings/Keybindings_en.md#:~:text=%60%60-,Cycle%20pagers,-%3Cesc%3E

Since the file is autogenerated, I guess there's sth wrong w the source code itself

**Comment by stefanhaller** [comment]
> In the docs it's mentioned that we can cycle through pagers with `|` key. But I can't find this shortcut anywhere!

This feature is not released yet, you'll need a build from master for now. The next release will be on Saturday (probably).


> I looked in the keybindings docs, there's an entry for "Cycle pagers" but the associated bind is empty: 

This looks like a bug with how the keybindings file is generated, I'll look into it when I get home this weekend.

**Comment by stefanhaller** [comment]
@gwennlbh The rendering problem in the keybindings file is fixed now, see #5007.

**Comment by van-sprundel** [comment]
For anyone stumbling on this issue, see https://github.com/jesseduffield/lazygit/blob/master/docs/Custom_Pagers.md to setup difft (or any other custom pager) :)
